### ISSUE TITLE
Missing tooltips in SYSTEM -> Einstellungen

### ISSUE THREAD

**Issue by jochi44** [contao/contao#5043]
### Affected version(s)

5.0.0-RC1

### Description

Some tooltips are missing in SYSTEM -> Einstellungen:

- Der Datumsformat-String wird mit der PHP-Funktion date() geparst.
- Bitte wählen Sie die Zeitzone des Servers aus.
- List der erlaubten HTML-Attribute, die nicht entfernt werden. Das Tag bzw. der Attributname * steht für alle Tags bzw. Attribute. Für Attribute mit Bindestrichen können Platzhalter wie z. B. data-* benutzt werden.
- Bitte legen Sie die Standardzugriffsrechte für Seiten und Artikel fest.

I'm happy to help if somebody tells me how to do it correctly.

**Comment by fritzmg** [comment]
<img src="https://user-images.githubusercontent.com/4970961/180948080-82abb8fc-1264-4b7a-8e51-a086ac4c7b56.png" width="360">

<img src="https://user-images.githubusercontent.com/4970961/180948440-cfb453ed-2551-42b2-b3b1-7031449982d2.png" width="280">

I cannot confirm the problem.

**Comment by jochi44** [comment]
I tested this again.
It's working in Firefox (102.0.1 (64-Bit)) and not working in Chrome (Version 103.0.5060.134 (Offizieller Build) (64-Bit)).
I'm running Windows 11 64 bit.

**Comment by fritzmg** [comment]
I still cannot confirm:

<img src="https://user-images.githubusercontent.com/4970961/180953162-29452f3a-a0a1-49f8-b414-58fbaacec6fb.png" width="360">

This screenshot ist from Chrome 103.0.5060.134 (64-bit) under Windows 10 x64 21H2 build 19044.1826.

**Comment by jochi44** [comment]
This is very strange.
On a 4.13.6 installation the tooltips are shown (same PC, Chrome 103.0.5060.134 (64-bit).

**Comment by aschempp** [comment]
is there any console error in your browser? We did update the tips implementation, so I it _could_ be a JS or cache problem.

**Comment by jochi44** [comment]
> is there any console error in your browser? We did update the tips implementation, so I it _could_ be a JS or cache problem.

Unfortunately no errors in the dev console of the browser.
Working on Firefox 103.0.1 (64-Bit)
Not working on Google Chrome 104.0.5112.81 (Offizieller Build) (64-Bit)

**Comment by Xirdion** [comment]
I can confirm the behavior especially for the time zone, the allowed HTML attributes and the access rights.
Unfortunately there are no log entries in the browser console.

I have added a debugger-statement within the tips.min.js file right at the beginning of the setup-function.
So for every element of the page it will run into the breakpoint.
So I came to this image where you can see, that the help text for the timezone is not there. It will be shown when the next section is added to the screen, but then the tooltip-event is no added to it.
![image](https://user-images.githubusercontent.com/17783055/186178030-2682a328-02aa-4217-9e32-bc5bb26793f5.png)

**Comment by leofeyer** [comment]
I can confirm the issue in Chrome 104 on macOS 12.5, however, only for the **Time zone** and **Allowed HTML attributes** fields. The other tooltips are working.

**Comment by ausi** [comment]
Does it appear with the debug mode as well as without?
Does it make a difference if you clear the cache (`rm -rf var/cache/`)?

If the cache makes a difference it might be a similar problem as https://github.com/contao/contao/issues/5029#issuecomment-1216531922

**Comment by leofeyer** [comment]
Nothing makes any difference. `prod` or `dev`, with cache or without – the tooltip of the **Time zone** field does not work in Chrome. (It does work in Safari and Firefox.)

**Comment by m-vo** [comment]
What causes this issue is that the query selector used in `setup()` does not match the element (`p.tl_tip`) if the passed `node` is the `p` itself. And this likely happens more often for the timezone select with lots of option elements. The other smaller widgets are more likely to be together in one chunk when streaming. 

To work around this, the code could be refactored as following:

**before:**
```js
node.querySelectorAll('p.tl_tip').forEach(function (el) {
    init(el, 0, 23, true);
});
```

**after:**
```js
const tips = node.matches('p.tl_tip') ? 
    [node] : node.querySelectorAll('p.tl_tip');

tips.forEach(function (el) {
    init(el, 0, 23, true);
});
```

Note, that you would also have to pass ` setup(document.documentElement);` then for the initialization bit. 

I actually did something similar for the `_table` component:
https://github.com/contao/contao/blob/596ea88104c4543adcdebe494064f77f4bd5f5df/core-bundle/contao/templates/_new/component/_table.html.twig#L148-L157

But instead of fixing all places with this usage, couldn't we build a reusable function/module with a nice promise based interface that does all these subtle things for you and just offers an API like `elementReady(<selector>).then(…)` or the like?

**Comment by m-vo** [comment]
> But instead of fixing all places with this usage, couldn't we build a reusable function/module with a nice promise based interface that does all these subtle things for you and just offers an API like elementReady(<el>).then(…) or the like?

An other thing that could be included in such a solution, is only dispatching what is currently in `init()` the first time. For elements higher up in the DOM, this will get called  *a lot*. I know that the current code in `tips.js` caters for this, but again something that should probably be abstracted, because it is quite important performance wise.

**Comment by ausi** [comment]
I wrote something similar some time ago that does exactly that:

```js
const subscribers = {};
const executedCallbacks = new WeakMap();

/**
 * @callback observeCallback
 * @param {HTMLElement} element
 * @param {boolean} firstCall
 */
/**
 * @param {string} selector
 * @param {observeCallback} callback
 */
export default function observeElement(selector, callback) {
  subscribers[selector] = subscribers[selector] || [];
  subscribers[selector].push(callback);

  // Observer already existing elements
  document.querySelectorAll(selector).forEach((element) => executeCallback(callback, element));
}

function executeCallback(callback, element) {
  if (!executedCallbacks.has(element)) {
    executedCallbacks.set(element, new WeakMap());
  }

  const firstCall = !executedCallbacks.get(element).has(callback);
  executedCallbacks.get(element).set(callback, true);

  // Queue a microtask to prevent failing callbacks from stopping further execution
  Promise.resolve(true).then(() => callback(element, firstCall));
}

const observer = new MutationObserver((mutationsList) => {
  const allSelectors = Object.keys(subscribers).join(',');
  const allElements = new Set();

  for (const mutation of mutationsList) {
    for (const node of mutation.addedNodes || []) {
      if (node.matches && node.matches(allSelectors)) {
        allElements.add(node);
      }
      if (node.querySelectorAll) {
        node.querySelectorAll(allSelectors).forEach((element) => allElements.add(element));
      }
    }
  }

  for (const element of allElements) {
    for (const [selector, callbacks] of Object.entries(subscribers)) {
      if (element.matches(selector)) {
        callbacks.forEach((callback) => executeCallback(callback, element));
      }
    }
  }
});

observer.observe(document.documentElement, {
  attributes: false,
  childList: true,
  subtree: true,
});
```

But it is written as a module and has state, so not sure how we can integrate something like this here.

**Comment by leofeyer** [comment]
@m-vo @ausi @aschempp How do we proceed here?

**Comment by m-vo** [comment]
As we cannot use stimulus controllers for this, I'd implement the fix as outlined in https://github.com/contao/contao/issues/5043#issuecomment-1236103398.

**Comment by aschempp** [comment]
I think https://github.com/terminal42/contao-conditionalformfields/commit/714e7e86a34a3f40659d6b0597dce40ce6bf7ea7 would be the correct fix. We should not initialize the whole script until the DOM itself is ready, only after that we should also add the mutation observer.

**Comment by fritzmg** [comment]
If you do it on domready, you won't really need a mutation observer anymore though in this case.

**Comment by aschempp** [comment]
yes you do, e.g. for tips loaded from subpalettes

**Comment by leofeyer** [comment]
This should have been fixed in Contao 5.1. @jochi44 Can you please verify?

**Comment by jochi44** [comment]
I can confirm that this is fixed in Contao 5.1.
